File: my_app/app/src/main/java/com/example/my_app/login/RegisterActivity.java
package com.example.my_app.login;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import android.content.Intent;
import android.os.Bundle;
import android.text.TextUtils;
import android.view.View;
import android.widget.AdapterView;
import android.widget.Button;
import android.widget.EditText;
import android.widget.Spinner;
import android.widget.TextView;
import android.widget.Toast;

import com.example.my_app.R;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.auth.AuthResult;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.firestore.FirebaseFirestore;

public class RegisterActivity extends AppCompatActivity {

    private EditText email, password;
    private Spinner accountType;
    private Button btn_register;
    private TextView link_to_login;
    private FirebaseAuth mAuth;
    private FirebaseFirestore mFirestore;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_register);
        email = findViewById(R.id.email);
        password = findViewById(R.id.password);
        accountType = findViewById(R.id.accountType);
        btn_register = findViewById(R.id.btnRegister);
        link_to_login = findViewById(R.id.tvLinkToLogin);

        btn_register.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                String user = email.getText().toString().trim();
                String pass = password.getText().toString().trim();

                if(TextUtils.isEmpty(user)) {
                    email.setError("You must enter a username.");
                    return;
                }
                else if(TextUtils.isEmpty(pass)) {
                    password.setError("You must enter password.");
                    return;
                }
                else if (pass.length() < 6) {
                    password.setError("Your password must be at least 6 characters.");
                    return;
                }

                mAuth = FirebaseAuth.getInstance();
                mAuth.createUserWithEmailAndPassword(email.getText().toString(), password.getText().toString())
                        .addOnCompleteListener(RegisterActivity.this, new OnCompleteListener<AuthResult>() {
                            @Override
                            public void onComplete(@NonNull Task<AuthResult> task) {
                                if (task.isSuccessful()) {
                                    // Sign in success, update UI with the signed-in user's information
                                    //Log.d(TAG, "createUserWithEmail:success");
                                    mFirestore = FirebaseFirestore.getInstance();
                                    mFirestore.collection("userTypes").add(new User(email.getText().toString(), accountType.getSelectedItem().toString()));
                                    Toast.makeText(RegisterActivity.this, "Register success!", Toast.LENGTH_SHORT).show();
                                    FirebaseUser user = mAuth.getCurrentUser();
                                    mAuth.signOut();
                                    updateUI(user);
                                } else {
                                    // If sign in fails, display a message to the user.
                                    //Log.w(TAG, "createUserWithEmail:failure", task.getException());
                                    Toast.makeText(RegisterActivity.this, "Register failed!",
                                            Toast.LENGTH_SHORT).show();
                                    mAuth.signOut();
                                    updateUI(null);
                                }

                                // ...
                            }
                        });
            }
        });
        link_to_login.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Intent intent = new Intent(RegisterActivity.this, LoginActivity.class);
                startActivity(intent);
            }
        });
        accountType.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> adapterView, View view, int i, long l) {

            }

            @Override
            public void onNothingSelected(AdapterView<?> adapterView) {

            }
        });
    }

    private void updateUI(Object user) {
        if (user != null) {
            Intent intent = new Intent(this, LoginActivity.class);
            startActivity(intent);
        }
    }
}

File: BK_FoodCourt/app/src/main/java/com/example/bk_foodcourt/RegisterActivity.kt
package com.example.bk_foodcourt

import android.content.Intent
import android.os.Bundle
import android.text.TextUtils
import android.view.View
import android.widget.*
import androidx.appcompat.app.AppCompatActivity
import com.example.bk_foodcourt.login.LoginActivity
import com.google.firebase.auth.FirebaseAuth

class RegisterActivity : AppCompatActivity() {
    var mUsername: EditText? = null
    var mPassword: EditText? = null
    var mConfirmPW: EditText? = null
    var mFullname: EditText? = null
    var mPhonenumber: EditText? = null
    var mAddress: EditText? = null
    var mBirthday: EditText? = null
    var mGender: RadioGroup? = null
    var mMale: RadioButton? = null
    var mFemale: RadioButton? = null
    var mOther: RadioButton? = null
    var mButtonSignUp: Button? = null
    var fAuth: FirebaseAuth? = null
    var mLogin: TextView? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_register)

        // initialization
        mUsername = findViewById(R.id.etUsername)
        mPassword = findViewById(R.id.etPassword)
        mConfirmPW = findViewById(R.id.etConfirmPassword)
        mFullname = findViewById(R.id.etFullname)
        mPhonenumber = findViewById(R.id.etPhoneNumber)
        mAddress = findViewById(R.id.etAddress)
        mBirthday = findViewById(R.id.etBirthday)
        mGender = findViewById(R.id.rgGender)
        mMale = findViewById(R.id.rbMale)
        mFemale = findViewById(R.id.rbFemale)
        mOther = findViewById(R.id.rbOther)
        mButtonSignUp = findViewById(R.id.btnSignUp)
        mLogin = findViewById(R.id.tvLogin)
        // authentication
        fAuth = FirebaseAuth.getInstance()

        // check Login or not
        if (fAuth!!.currentUser != null) {
            startActivity(Intent(applicationContext, OrderActivity::class.java))
            finish()
        }
        // Register button click
        mButtonSignUp!!.setOnClickListener(View.OnClickListener {
            val username = mUsername!!.getText().toString().trim { it <= ' ' }
            val password = mPassword!!.getText().toString().trim { it <= ' ' }
            val confirmPassword =
                mConfirmPW!!.getText().toString().trim { it <= ' ' }
            if (TextUtils.isEmpty(username)) {
                mUsername!!.setError("You must enter a username.")
                return@OnClickListener
            } else if (TextUtils.isEmpty(password)) {
                mPassword!!.setError("You must enter password.")
                return@OnClickListener
            } else if (password.length < 6) {
                mPassword!!.setError("Your password must be at least 6 characters.")
                return@OnClickListener
            } else if (password == confirmPassword == false) {
                mPassword!!.setError("Your password and confirm password don't match.")
                return@OnClickListener
            }


            // register the user in firebase
            fAuth!!.createUserWithEmailAndPassword(username, password)
                .addOnCompleteListener { task ->
                    if (task.isSuccessful) {
                        Toast.makeText(
                            this@RegisterActivity,
                            "Sign up successfully!",
                            Toast.LENGTH_SHORT
                        ).show()
                        fAuth!!.signOut()
                        startActivity(
                            Intent(
                                applicationContext,
                                MainActivity::class.java
                            )
                        )
                        finish()
                    } else {
                        Toast.makeText(
                            this@RegisterActivity,
                            "Error! " + task.exception!!.message,
                            Toast.LENGTH_SHORT
                        ).show()
                    }
                }
        })
        mLogin!!.setOnClickListener(View.OnClickListener {
            startActivity(Intent(applicationContext, LoginActivity::class.java))
            finish()
        })
    }
}